### ISSUE TITLE
bug(cheatcodes): differentiate between `FOO=` (unset) and `FOO=""` (empty) in `vm.envOr`

### ISSUE THREAD

**Issue by Ratimon** [foundry-rs/foundry#7408]
### Component

Forge

### Have you ensured that all of these are up to date?

- [ ] Foundry
- [ ] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (2d54c1f 2024-03-03T00:21:12.158864000Z)

### What command(s) is the bug in?

forge script

### Operating System

None

### Describe the bug

Reference

My attempts to call `writeJson` cheat code from the following snippet: 

<img width="854" alt="Screenshot 2567-03-15 at 15 10 00" src="https://github.com/foundry-rs/foundry/assets/22450628/ce04843f-64e1-4202-85d6-a3c40ebbc0b3">

[`It is called from this snippet`](https://github.com/Ratimon/optimism-contracts-foundry/blob/9cd1a8d3761f9e3d053a5fac6653f0371130d963/script/deployer/Deployer.sol#L333C1-L337C1)

I note that if I hardcode the env var of `DEPLOYMENT_OUTFILE=/deployments/.deploy` in my  [`.env.local`](https://github.com/Ratimon/optimism-contracts-foundry/blob/9cd1a8d3761f9e3d053a5fac6653f0371130d963/.env) , it works. However when I tried to fallback as following, it suddenly fails as shown above
<img width="917" alt="Screenshot 2567-03-15 at 15 09 13" src="https://github.com/foundry-rs/foundry/assets/22450628/8d83c5a3-3d05-4dc2-8b7b-b9b274139b29">

[Config file](https://github.com/Ratimon/optimism-contracts-foundry/blob/9cd1a8d3761f9e3d053a5fac6653f0371130d963/script/deployer/Config.sol)

For further reference, this is my foundry config and I have set the write access.
<img width="510" alt="Screenshot 2567-03-15 at 15 10 41" src="https://github.com/foundry-rs/foundry/assets/22450628/ccb099e6-7e11-4ea7-b366-39649ea47ddb">

[Foundry config file](https://github.com/Ratimon/optimism-contracts-foundry/blob/9cd1a8d3761f9e3d053a5fac6653f0371130d963/foundry.toml)

All source code can be found here: [code](https://github.com/Ratimon/optimism-contracts-foundry/commit/9cd1a8d3761f9e3d053a5fac6653f0371130d963)

**Comment by klkvr** [comment]
We are currently explicitly disallow writing to foundry.toml: https://github.com/foundry-rs/foundry/blob/3fa02706ca732c994715ba42d923605692062375/crates/cheatcodes/src/config.rs#L126

so this is expected, not sure if we consider it reasonable to remove this check at this point cc @mds1 @mattsse

**Comment by mattsse** [comment]
yeah we don't allow this, because this could be used to inject settings such as permissions, however for the current run the config is already loaded and would be unaffected by this.

what's your use case for this @Ratimon ?

**Comment by Ratimon** [comment]
> yeah we don't allow this, because this could be used to inject settings such as permissions, however for the current run the config is already loaded and would be unaffected by this.
> 
> what's your use case for this @Ratimon ?

I am sorry for not writing clear issue at first. What I encountered is that I can not access to other folders/directory even though I have set the permission access in foundry.toml

In my case, it is `/deployments/` where I set to `read-write` as shown aove.

**Comment by mattsse** [comment]
ah okay, so this is actually a bug because you don't actually target the foundry.toml file?

could you please add the test as plaintext here so we can try to replicate the issue?

**Comment by Ratimon** [comment]
> ah okay, so this is actually a bug because you don't actually target the foundry.toml file?
> 
> could you please add the test as plaintext here so we can try to replicate the issue?

Sure!!

**Repro Steps**

1) git clone -b reproduce-bug git@github.com:Ratimon/optimism-contracts-foundry.git

2) install dependencies
```bash
pnpm i
```

2) run `pnpm deploy_000_safe`

I Note that I target [.deploy.json] (https://github.com/Ratimon/optimism-contracts-foundry/blob/main/deployments/31337/.deploy.json)

**Comment by klkvr** [comment]
@Ratimon from your trace I see that `vm.writeJson` is getting called with empty "path" parameter, any chance `DEPLOYMENT_OUTFILE` might be an empty variable in your environment?

This check is getting hit because it's being performed via `.starts_with` and any path starts with an empty path

**Comment by Ratimon** [comment]
> @Ratimon from your trace I see that `vm.writeJson` is getting called with empty "path" parameter, any chance `DEPLOYMENT_OUTFILE` might be an empty variable in your environment?
> 
> This check is getting hit because it's being performed via `.starts_with` and any path starts with an empty path

Yes, `DEPLOYMENT_OUTFILE` is empty and it should be fallback to this [Config file](https://github.com/Ratimon/optimism-contracts-foundry/blob/9cd1a8d3761f9e3d053a5fac6653f0371130d963/script/deployer/Config.sol) as shown below:

<img width="917" alt="Screenshot 2567-03-15 at 15 09 13" src="https://github.com/foundry-rs/foundry/assets/22450628/8d83c5a3-3d05-4dc2-8b7b-b9b274139b29">

**Comment by klkvr** [comment]
So the issue here is that we treat `DEPLOYMENT_OUTFILE=` entry in .env file the same way as `DEPLOYMENT_OUTFILE=""`, thus we don't actually fallback to the provided default. Not sure if this is expected, but I am pretty sure that there are usecases for env vars with empty strings

You can try removing DEPLOYMENT_OUTFILE line from .env.local at all and this should get fixed

**Comment by Ratimon** [comment]
> So the issue here is that we treat `DEPLOYMENT_OUTFILE=` entry in .env file the same way as `DEPLOYMENT_OUTFILE=""`, thus we don't actually fallback to the provided default. Not sure if this is expected, but I am pretty sure that there are usecases for env vars with empty strings
> 
> You can try removing DEPLOYMENT_OUTFILE line from .env.local at all and this should get fixed

Yes!! I tried removing and my issue is resolved!! Thanks!!

Personally, I think that env vars with empty strings are important for code readability in smart contract project.

**Comment by zerosnacks** [comment]
Converting this issue into an actionable ticket focused on differentiating between unset variables (`FOO=`) and empty string variables (`FOO=""`)

**Comment by DaniPopes** [comment]
fyi both FOO= and FOO="" are set with empty value, there is no syntax for unset

**Comment by DaniPopes** [comment]
We decided not to implement this as the empty string is a valid value. You can unset variables in the console with `unset VAR` or not set them in the first place by commenting the line with `#`.
